aiida/orm/implementation/sqlalchemy/querybuilder/main.py

```diff
@@ -169,117 +202,75 @@ def get_session(self) -> Session:
         return self._backend.get_session()
 
     def count(self, data: QueryDictType) -> int:
-        query = self.update_query(data)
-        try:
-            return query.count()
-        except Exception:
-            self.get_session().close()
-            raise
-
-    def yield_per(self, data: QueryDictType, batch_size):
-        query = self.update_query(data)
-        try:
-            return query.yield_per(batch_size)
-        except Exception:
-            self.get_session().close()
-            raise
+        with self.use_query(data) as query:
+            result = query.count()
+        return result
 
     def first(self, data: QueryDictType) -> Optional[List[Any]]:
-        query = self.update_query(data)
-        try:
+        with self.use_query(data) as query:
             result = query.first()
-        except Exception:
-            self.get_session().close()
-            raise
 
         if result is None:
             return result
 
-        if not isinstance(result, (list, tuple)):
-            result = [result]
+        # we discard the first item of the result row,
+        # which was what the query was initialised with and not one of the requested projection (see self._build)
+        result = result[1:]
 
-        if not self._projection_index_to_field or len(result) != len(self._projection_index_to_field):
-            raise Exception('length of query result does not match the number of specified projections')
+        if len(result) != self._requested_projections:
+            raise Exception(
```

Comment:
Should probably use a more specific exception and not the base class. Would this be an internal coding error? If so, probably `AssertionError` would be best. And in that case change the conditional into simple `assert len(result) == self._requested_projections`.

Reply:
yep `AssertionError` sounds good.
I recently found out that if you run Python with the `-O` flag, it skips all `assert` statements, and thats why they are discouraged. No idea what the use case is to run with that though

Reply:
done